--- git/Makefile	2018-03-20 15:14:37.542656133 -0400
+++ git/Makefile	2018-03-20 15:26:49.154234379 -0400
@@ -125,7 +125,7 @@
 	$(INSTALL_X) bin/turbovisor $(INSTALL_BIN)
 	@echo "==== Building 3rdparty modules ===="
 	make -C deps/http-parser library
-	$(CC) $(INC) -shared -O3 -Wall $(CFLAGS) $(HTTP_PARSERDIR)/libhttp_parser.o $(TDEPS)/turbo_ffi_wrap.c -o $(INSTALL_TFFI_WRAP_SOSHORT) $(LDFLAGS)
+	$(CC) $(INC) -shared -O3 -Wall -Wl,-soname,$(INSTALL_TFFI_WRAP_SOSHORT) $(CFLAGS) $(HTTP_PARSERDIR)/libhttp_parser.o $(TDEPS)/turbo_ffi_wrap.c -o $(INSTALL_TFFI_WRAP_SOSHORT) $(LDFLAGS)
 	@echo "==== Installing libtffi_wrap ===="
 ifeq ($(uname_S),Linux)
 	test -f $(INSTALL_TFFI_WRAP_SOSHORT) && \
